### ISSUE TITLE
[Question] handle not all charts being applied in all environments

### ISSUE THREAD

**Issue by valdisrigdon** [roboll/helmfile#482]
I have a use case of running 3+ clusters and am trying to manage the configuration of all of them. I've organized the charts by team/service and am using an `environment` per cluster.  What I'm trying to do is support a structure where each team can declare the charts they want deployed to any cluster.

My structure looks like this, with an example of 3 clusters:

- `helmfile.yaml`
  - `projects/`
    - `serviceA`
      - `cluster1.yaml`
      - `cluster2.yaml`
    - `serviceB`
      - `cluster1.yaml`
      - `cluster3.yaml`

`helmfile.yaml` looks something like this:

```yaml
environments:
- cluster1:
- cluster2:
- cluster3:

helmfiles:
- projects/*.yaml
```

and each `clusterN.yaml` file in the `serviceX` includes an `environments` section that lists that particular `environment`, so `serviceA/cluster2.yaml` looks like:

```yaml
environments:
 - cluster2

releases:
- chart: stable/grafana
- name: grafana
...
```

When running `helmfile -f projects/helmfile.yaml -e cluster2` apply, `serviceA/cluster1.yaml` will be included by the `helmfiles` pattern in `helmfile.yaml` and I see in the output something like:

```
could not deduce `environment:` block, configuring only .Environment.Name. error: failed to read cluster1.yaml: environment "cluster1" is not defined
```
The correct actions are being applied, but it's troublesome that I'm seeing these error messages and I'm wondering if there is a better way to structure these files to avoid the error. The code as written seems to indicate that my setup is an error.  If it's a supported use case, then I can open a PR to change the log level and warning.

To summarize the use case, we support 15+ clusters, and each team will deploy a set of charts to some of these clusters, and those charts aren't the same for each cluster. Some teams only ever deploy to a subset of these clusters.  I also want to be able to create separate files for different clusters instead of use templating; it's easier to read when trying to understand what charts are deployed to which cluster.  I also like the file structure where each team can organize their charts in a `serviceX` folder instead of grouping by cluster.

**Comment by valdisrigdon** [comment]
@mumoshu Any ideas here?

**Comment by mumoshu** [comment]
@valdisrigdon Hey! I'm still trying to understand, but the error message is there to notice you about the potential mistake that you missed defining the environment for not all but a subset of helmfiles, although it isn't the case this time.

Could you, by any chance, anything other than `cluster` as the unit of environments?

I usually define environments like `prod`, `staging`, and `dev` and all the sub-helmfiles includes the three environments in their `environments:` section. In my case, for example, cluster1 to cluster3 may belong to the `prod` environment, and cluster4 to cluster5 may belong to the `staging` environment.

This way, I don't get bothered when someone creates a brand-new cluster(say, cluster6) and service owners has not yet created the helmfile environment for the cluster6.

It is the cluster owner's responsibility to choose the appropriate helmfile environment for the cluster.

Please feel free to leave more comments if I'm missing the point.

**Comment by valdisrigdon** [comment]
Not all charts are applied to all clusters/environments.  In our case, the 15+ clusters support a variety of workloads and they don't all fit into a `dev` or `prod` type environment.

In other words, we like to organize by team/project, but then apply by cluster.  We don't care which cluster that each team/project deploys their services to, so that's not limited in anyway. The only way I found to make that happen is to list *all* possible environments in the top-level `helmfile.yaml` because if I run `helmfile -e cluster helmfile.yaml` I need `cluster1` as a listed environment in `helmfile.yaml`.  Would a possible change be to continue processing the referenced `helmfiles` even if a `-e` environment isn't listed there?

**Comment by mumoshu** [comment]
@valdisrigdon Thanks for the explanation! I think I understood your problem. Sounds very valid.

How should we make it configurable? My quick idea is that we add a new option `missingEnvironmentHandler: Info` in the top-level of each helmfile.yaml, that works like `releases[].missingFileHandler` added in #439.
Does that work for you?
